runtime/parachains/src/router/ump.rs

```diff
@@ -41,32 +60,582 @@ impl<T: Trait> Module<T> {
+			QueueCacheEntry {
+				queue,
+				count,
+				total_size,
+			}
+		});
+		let upward_message = cache_entry.queue.pop_front();
+		if let Some(ref msg) = upward_message {
+			cache_entry.count -= 1;
+			cache_entry.total_size -= msg.len() as u32;
+		}
+
+		let became_empty = cache_entry.queue.is_empty();
+		(upward_message, became_empty)
+	}
+
+	/// Flushes the updated queues into the storage.
+	fn flush<T: Trait>(self) {
+		for (
+			para,
+			QueueCacheEntry {
+				queue,
+				count,
+				total_size,
+			},
+		) in self.0
+		{
+			if queue.is_empty() {
+				// remove the entries altogether.
+				<Module<T> as Store>::RelayDispatchQueues::remove(&para);
+				<Module<T> as Store>::RelayDispatchQueueSize::remove(&para);
+			} else {
+				<Module<T> as Store>::RelayDispatchQueues::insert(&para, queue);
+				<Module<T> as Store>::RelayDispatchQueueSize::insert(&para, (count, total_size));
+			}
+		}
+	}
+}
+
+/// A cursor that iterates over all entries in `NeedsDispatch`.
+///
+/// This cursor will start with the para indicated by `NextDispatchRoundStartWith` storage entry.
+/// This cursor is cyclic meaning that after reaching the end it will jump to the beginning. Unlike
+/// an iterator, this cursor allows removing items during the iteration.
+///
+/// Each iteration cycle *must be* concluded with a call to either `advance` or `remove`.
+///
+/// This struct is not supposed to be dropped but rather to be consumed by [`flush`].
+#[derive(Debug)]
+struct NeedsDispatchCursor {
+	needs_dispatch: Vec<ParaId>,
+	cur_idx: usize,
+}
+
+impl NeedsDispatchCursor {
+	fn new<T: Trait>() -> Self {
+		let needs_dispatch: Vec<ParaId> = <Module<T> as Store>::NeedsDispatch::get();
+		let start_with = <Module<T> as Store>::NextDispatchRoundStartWith::get();
+
+		let start_with_idx = match start_with {
+			Some(para) => match needs_dispatch.binary_search(&para) {
+				Ok(found_idx) => found_idx,
+				Err(_supposed_idx) => {
+					// well that's weird because we maintain an invariant that
+					// `NextDispatchRoundStartWith` must point into one of the items in
+					// `NeedsDispatch`.
+					//
+					// let's select 0 as the starting index as a safe bet.
+					debug_assert!(false);
+					0
+				}
+			},
+			None => 0,
+		};
+
+		Self {
+			needs_dispatch,
+			cur_idx: start_with_idx,
+		}
+	}
+
+	/// Returns the item the cursor points to.
+	fn peek(&self) -> Option<ParaId> {
+		self.needs_dispatch.get(self.cur_idx).cloned()
+	}
+
+	/// Moves the cursor to the next item.
+	fn advance(&mut self) {
+		if self.needs_dispatch.is_empty() {
+			return;
+		}
+		self.cur_idx = (self.cur_idx + 1) % self.needs_dispatch.len();
+	}
+
+	/// Removes the item under the cursor.
+	fn remove(&mut self) {
+		if self.needs_dispatch.is_empty() {
+			return;
+		}
+		let _ = self.needs_dispatch.remove(self.cur_idx);
+	}
+
+	/// Flushes the dispatcher state into the persistent storage.
+	fn flush<T: Trait>(self) {
```

Comment:
Would it be useful to have a `Drop` impl which calls `self.flush()`? That's the behavior I expect from `Write` trait instances, at least.

Reply:
While I understand that it's more idiomatic to lean on RAII generally, runtime code is exception, I'd say. The reason is, when the runtime is used in native mode, i.e. it's linked against the main binary as a normal rust crate, we use `panic=unwind` there to handle potential panics.

A big caveat with that, if this drop impl was invoked during panicking within runtime then that would lead to a double panic. And double panic leads to a process abort which is a disaster. Because of that, and because the code here is local and used only in a function with straghtforward control flow, I thought I'd go to an explicit function.

That said, I thought I had left a comment about this but apparently I didn't. I will leave one